File: scripts/clean_cedar_systems_json.py
#!/usr/bin/env python3
"""
Clean CEDAR system summary JSON exports corrupted by pasted ISO timestamps.

Some exports accidentally insert timestamps into the middle of keys and values, e.g.
  "ictObjectI2026-03-31T16:04:35.409172087Z d" instead of "ictObjectId"
  "Health Plan2026-03-31T16:04:35.409172087Z Management System"

Usage:
  python3 scripts/clean_cedar_systems_json.py path/to/systems.json
  python3 scripts/clean_cedar_systems_json.py impl.json prod.json -o cleaned/
  python3 scripts/clean_cedar_systems_json.py systems.json --in-place
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path
from typing import Any

# Timestamps accidentally pasted during copy/export (any ISO-8601 UTC with fractional seconds).
TIMESTAMP_PATTERN = re.compile(r"20\d{2}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+Z")

# Spaces left in UUIDs when a timestamp sat between hex segments (e.g. "3193C9 0E117E").
UUID_SEGMENT_GAP = re.compile(
    r"(\{?[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]*)"
    r" ([0-9A-Fa-f]{4,}\}?)"
)


def clean_string(value: str) -> tuple[str, bool]:
    """Remove timestamp fragments and repair UUID spacing. Returns (cleaned, changed)."""
    original = value
    cleaned = TIMESTAMP_PATTERN.sub("", value)
    prev = None
    while prev != cleaned:
        prev = cleaned
        cleaned = UUID_SEGMENT_GAP.sub(r"\1\2", cleaned)
    cleaned = re.sub(r"(-) +", r"\1", cleaned)
    cleaned = re.sub(r" +(-)", r"\1", cleaned)
    return cleaned, cleaned != original


def clean_system_object(obj: dict[str, Any]) -> tuple[dict[str, Any], list[str]]:
    """Return a cleaned system object and human-readable change notes."""
    changes: list[str] = []
    cleaned: dict[str, Any] = {}

    for key, value in obj.items():
        new_key, key_changed = clean_string(key)
        if key_changed:
            changes.append(f"key {key!r} -> {new_key!r}")

        if isinstance(value, str):
            new_value, value_changed = clean_string(value)
            if value_changed:
                label = new_key if not key_changed else f"{key!r} (field)"
                preview = value if len(value) <= 60 else value[:57] + "..."
                changes.append(f"{label}: removed timestamp from value ({preview!r})")
        else:
            new_value = value

        if new_key in cleaned and cleaned[new_key] != new_value:
            raise ValueError(
                f"duplicate key after cleaning: {new_key!r} "
                f"(existing {cleaned[new_key]!r}, new {new_value!r})"
            )
        cleaned[new_key] = new_value

    return cleaned, changes


def load_systems(path: Path) -> list[dict[str, Any]]:
    with path.open(encoding="utf-8") as handle:
        data = json.load(handle)
    if not isinstance(data, list):
        raise ValueError(f"{path}: expected a JSON array of system objects")
    for index, item in enumerate(data):
        if not isinstance(item, dict):
            raise ValueError(f"{path}: item {index} is not an object")
    return data


def audit(data: list[dict[str, Any]]) -> list[str]:
    """Return descriptions of remaining timestamp corruption."""
    issues: list[str] = []
    for index, obj in enumerate(data):
        for key, value in obj.items():
            if TIMESTAMP_PATTERN.search(key):
                issues.append(f"[{index}] key still has timestamp: {key!r}")
            if isinstance(value, str) and TIMESTAMP_PATTERN.search(value):
                system = obj.get("acronym") or obj.get("name") or obj.get("id")
                issues.append(f"[{index}] {system}: field {key!r} still has timestamp")
    return issues


def write_systems(path: Path, data: list[dict[str, Any]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8") as handle:
        json.dump(data, handle, indent=2, ensure_ascii=False)
        handle.write("\n")


def process_file(
    input_path: Path,
    output_path: Path,
    *,
    dry_run: bool,
    verbose: bool,
) -> int:
    systems = load_systems(input_path)
    cleaned_systems: list[dict[str, Any]] = []
    systems_with_changes = 0
    all_changes: list[str] = []

    for index, system in enumerate(systems):
        cleaned, changes = clean_system_object(system)
        cleaned_systems.append(cleaned)
        if changes:
            systems_with_changes += 1
            acronym = system.get("acronym") or system.get("name") or f"index {index}"
            for change in changes:
                all_changes.append(f"{acronym}: {change}")

    remaining = audit(cleaned_systems)

    if verbose or dry_run:
        print(f"{input_path}:", file=sys.stderr)
        print(f"  systems: {len(systems)}", file=sys.stderr)
        print(f"  systems with fixes: {systems_with_changes}", file=sys.stderr)
        if verbose and all_changes:
            for note in all_changes:
                print(f"  - {note}", file=sys.stderr)
        if remaining:
            print(f"  remaining issues: {len(remaining)}", file=sys.stderr)
            for issue in remaining:
                print(f"    ! {issue}", file=sys.stderr)

    if remaining:
        print(f"error: {input_path} still has corruption after cleaning", file=sys.stderr)
        return 1

    if dry_run:
        print(f"dry run: would write {output_path}", file=sys.stderr)
        return 0

    write_systems(output_path, cleaned_systems)
    print(f"wrote {output_path}", file=sys.stderr)
    return 0


def resolve_output_path(input_path: Path, output_dir: Path | None, in_place: bool) -> Path:
    if in_place:
        return input_path
    if output_dir is not None:
        return output_dir / input_path.name
    return input_path.with_name(input_path.stem + "_cleaned" + input_path.suffix)


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Remove pasted ISO timestamps from CEDAR systems JSON exports.",
    )
    parser.add_argument(
        "inputs",
        nargs="+",
        type=Path,
        help="One or more JSON files (array of system summary objects)",
    )
    parser.add_argument(
        "-o",
        "--output-dir",
        type=Path,
        help="Directory for cleaned files (default: <name>_cleaned.json next to input)",
    )
    parser.add_argument(
        "--in-place",
        action="store_true",
        help="Overwrite input files instead of writing new paths",
    )
    parser.add_argument(
        "-n",
        "--dry-run",
        action="store_true",
        help="Report changes without writing files",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Print per-field fix details",
    )
    args = parser.parse_args()

    if args.in_place and args.output_dir is not None:
        print("error: use either --in-place or --output-dir, not both", file=sys.stderr)
        return 2

    exit_code = 0
    for input_path in args.inputs:
        if not input_path.is_file():
            print(f"error: not a file: {input_path}", file=sys.stderr)
            exit_code = 1
            continue
        output_path = resolve_output_path(input_path, args.output_dir, args.in_place)
        exit_code = max(exit_code, process_file(
            input_path,
            output_path,
            dry_run=args.dry_run,
            verbose=args.verbose,
        ))

    return exit_code


if __name__ == "__main__":
    sys.exit(main())

File: scripts/clean-cedar-systems-json.mjs
#!/usr/bin/env node

import { mkdir, readFile, stat, writeFile } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { inspect } from "node:util";

// Timestamps accidentally pasted during copy/export (any ISO-8601 UTC with fractional seconds).
const TIMESTAMP_PATTERN = /20\d{2}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+Z/g;

// Spaces left in UUIDs when a timestamp sat between hex segments (e.g. "3193C9 0E117E").
const UUID_SEGMENT_GAP =
  /(\{?[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]*) ([0-9A-Fa-f]{4,}\}?)/g;
const WRAPPED_UUID =
  /^\{+([0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12})\}+$/;

async function main() {
  const args = parseArgs(process.argv.slice(2));

  if (args.inPlace && args.outputDir !== null) {
    console.error("error: use either --in-place or --output-dir, not both");
    process.exitCode = 2;
    return;
  }

  let exitCode = 0;

  for (const inputPath of args.inputs) {
    if (!await isFile(inputPath)) {
      console.error(`error: not a file: ${inputPath}`);
      exitCode = Math.max(exitCode, 1);
      continue;
    }

    const outputPath = resolveOutputPath(inputPath, args.outputDir, args.inPlace);
    const result = await processFile(inputPath, outputPath, {
      dryRun: args.dryRun,
      verbose: args.verbose,
    });
    exitCode = Math.max(exitCode, result);
  }

  process.exitCode = exitCode;
}

function parseArgs(argv) {
  const args = {
    dryRun: false,
    inPlace: false,
    inputs: [],
    outputDir: null,
    verbose: false,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const current = argv[index];

    if (current === "--help" || current === "-h") {
      printHelp();
      process.exit(0);
    }

    if (current === "--in-place") {
      args.inPlace = true;
      continue;
    }

    if (current === "--dry-run" || current === "-n") {
      args.dryRun = true;
      continue;
    }

    if (current === "--verbose" || current === "-v") {
      args.verbose = true;
      continue;
    }

    if (current === "--output-dir" || current === "-o") {
      const outputDir = argv[index + 1];
      if (!outputDir || outputDir.startsWith("-")) {
        throw new Error(`${current} requires a directory path.`);
      }
      args.outputDir = outputDir;
      index += 1;
      continue;
    }

    if (!current.startsWith("-")) {
      args.inputs.push(current);
      continue;
    }

    throw new Error(`Unknown argument: ${current}`);
  }

  if (args.inputs.length === 0) {
    throw new Error("At least one JSON input file is required.");
  }

  return args;
}

function printHelp() {
  console.log(`Usage:
  node ./scripts/clean-cedar-systems-json.mjs path/to/systems.json
  node ./scripts/clean-cedar-systems-json.mjs impl.json prod.json -o cleaned/
  node ./scripts/clean-cedar-systems-json.mjs systems.json --in-place

Options:
  -o, --output-dir <dir>  Directory for cleaned files. Defaults to <name>_cleaned.json next to input.
  --in-place             Overwrite input files instead of writing new paths.
  -n, --dry-run          Report changes without writing files.
  -v, --verbose          Print per-field fix details.
  -h, --help             Show this help message.
`);
}

async function isFile(filePath) {
  try {
    return (await stat(filePath)).isFile();
  } catch (error) {
    if (error.code === "ENOENT") {
      return false;
    }
    throw error;
  }
}

async function processFile(inputPath, outputPath, { dryRun, verbose }) {
  const systems = await loadSystems(inputPath);
  const cleanedSystems = [];
  let systemsWithChanges = 0;
  const allChanges = [];

  for (const [index, system] of systems.entries()) {
    const { cleaned, changes } = cleanSystemObject(system);
    cleanedSystems.push(cleaned);

    if (changes.length > 0) {
      systemsWithChanges += 1;
      const acronym = system.acronym || system.name || `index ${index}`;
      for (const change of changes) {
        allChanges.push(`${acronym}: ${change}`);
      }
    }
  }

  const remaining = audit(cleanedSystems);

  if (verbose || dryRun) {
    console.error(`${inputPath}:`);
    console.error(`  systems: ${systems.length}`);
    console.error(`  systems with fixes: ${systemsWithChanges}`);

    if (verbose && allChanges.length > 0) {
      for (const note of allChanges) {
        console.error(`  - ${note}`);
      }
    }

    if (remaining.length > 0) {
      console.error(`  remaining issues: ${remaining.length}`);
      for (const issue of remaining) {
        console.error(`    ! ${issue}`);
      }
    }
  }

  if (remaining.length > 0) {
    console.error(`error: ${inputPath} still has corruption after cleaning`);
    return 1;
  }

  if (dryRun) {
    console.error(`dry run: would write ${outputPath}`);
    return 0;
  }

  await writeSystems(outputPath, cleanedSystems);
  console.error(`wrote ${outputPath}`);
  return 0;
}

async function loadSystems(filePath) {
  const rawJson = await readFile(filePath, "utf8");
  const data = JSON.parse(rawJson);

  if (!Array.isArray(data)) {
    throw new Error(`${filePath}: expected a JSON array of system objects`);
  }

  for (const [index, item] of data.entries()) {
    if (!isPlainObject(item)) {
      throw new Error(`${filePath}: item ${index} is not an object`);
    }
  }

  return data;
}

function isPlainObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function cleanSystemObject(obj) {
  const changes = [];
  const cleaned = {};

  for (const [key, value] of Object.entries(obj)) {
    const { value: newKey, changed: keyChanged } = cleanString(key, {
      compactSpaces: true,
    });

    if (keyChanged) {
      changes.push(`key ${formatValue(key)} -> ${formatValue(newKey)}`);
    }

    let newValue = value;
    if (typeof value === "string") {
      const result = cleanString(value);
      newValue = result.value;

      if (result.changed) {
        const label = keyChanged ? `${formatValue(key)} (field)` : newKey;
        const preview = value.length <= 60 ? value : `${value.slice(0, 57)}...`;
        changes.push(`${label}: cleaned value (${formatValue(preview)})`);
      }
    }

    if (
      Object.hasOwn(cleaned, newKey) &&
      JSON.stringify(cleaned[newKey]) !== JSON.stringify(newValue)
    ) {
      throw new Error(
        `duplicate key after cleaning: ${formatValue(newKey)} ` +
          `(existing ${formatValue(cleaned[newKey])}, new ${formatValue(newValue)})`,
      );
    }

    cleaned[newKey] = newValue;
  }

  return { cleaned, changes };
}

function cleanString(value, { compactSpaces = false } = {}) {
  const original = value;
  let cleaned = value.replace(TIMESTAMP_PATTERN, "");
  let previous = null;

  while (previous !== cleaned) {
    previous = cleaned;
    cleaned = cleaned.replace(UUID_SEGMENT_GAP, "$1$2");
  }

  cleaned = cleaned.replace(/(-) +/g, "$1");
  cleaned = cleaned.replace(/ +(-)/g, "$1");
  cleaned = cleaned.replace(WRAPPED_UUID, "$1");

  if (compactSpaces) {
    cleaned = cleaned.replace(/ +/g, "");
  }

  return {
    changed: cleaned !== original,
    value: cleaned,
  };
}

function audit(data) {
  const issues = [];

  for (const [index, obj] of data.entries()) {
    for (const [key, value] of Object.entries(obj)) {
      if (hasTimestamp(key)) {
        issues.push(`[${index}] key still has timestamp: ${formatValue(key)}`);
      }

      if (typeof value === "string" && hasTimestamp(value)) {
        const system = obj.acronym || obj.name || obj.id;
        issues.push(
          `[${index}] ${system}: field ${formatValue(key)} still has timestamp`,
        );
      }
    }
  }

  return issues;
}

function hasTimestamp(value) {
  TIMESTAMP_PATTERN.lastIndex = 0;
  return TIMESTAMP_PATTERN.test(value);
}

async function writeSystems(filePath, data) {
  await mkdir(path.dirname(filePath), { recursive: true });
  await writeFile(filePath, `${JSON.stringify(data, null, 2)}\n`, "utf8");
}

function resolveOutputPath(inputPath, outputDir, inPlace) {
  if (inPlace) {
    return inputPath;
  }

  if (outputDir !== null) {
    return path.join(outputDir, path.basename(inputPath));
  }

  const parsed = path.parse(inputPath);
  return path.join(parsed.dir, `${parsed.name}_cleaned${parsed.ext}`);
}

function formatValue(value) {
  return inspect(value, { breakLength: Infinity });
}

main().catch((error) => {
  console.error(error.message);
  process.exitCode = 1;
});
